indexing-service/src/main/java/org/apache/druid/indexing/scheduledbatch/ScheduledBatchTaskManager.java

```diff
@@ -0,0 +1,298 @@
+        }
+      }
+    };
+  }
+
+  /**
+   * Initializes and starts scheduled ingestion for the specified {@code supervisorId}, scheduling tasks
+   * using the provided {@code spec} and {@code cronSchedulerConfig}.
+   */
+  public void startScheduledIngestion(
+      final String supervisorId,
+      final String dataSource,
+      final CronSchedulerConfig cronSchedulerConfig,
+      final ClientSqlQuery spec
+  )
+  {
+    log.info(
+        "Starting scheduled batch ingestion for supervisorId[%s] with datasource[%s], cronSchedule[%s] and spec[%s].",
+        supervisorId, dataSource, cronSchedulerConfig, spec
+    );
+    final SchedulerManager manager = new SchedulerManager(supervisorId, dataSource, cronSchedulerConfig, spec);
+    manager.startScheduling();
+    supervisorToManager.put(supervisorId, manager);
+  }
+
+  /**
+   * Stops the scheduler for the specified {@code supervisorId}, pausing task submissions
+   * while retaining the ability to track the supervisor's status.
+   */
+  public void stopScheduledIngestion(final String supervisorId)
+  {
+    log.info("Stopping scheduled batch ingestion for supervisorId[%s].", supervisorId);
+    final SchedulerManager manager = supervisorToManager.get(supervisorId);
+    if (manager != null) {
+      // Don't remove the supervisorId from supervisorToManager. We want to be able to track the status
+      // for suspended supervisors to track any in-flight tasks. stop() will clean up all state completely.
+      manager.stopScheduling();
+    }
+  }
+
+  public void start()
+  {
+    log.info("Starting scheduled batch scheduler.");
+    final Optional<TaskRunner> taskRunnerOptional = taskMaster.getTaskRunner();
+    if (taskRunnerOptional.isPresent()) {
+      taskRunnerOptional.get().registerListener(taskRunnerListener, Execs.directExecutor());
+    } else {
+      log.warn("Task runner not registered for scheduled batch scheduler.");
+    }
+  }
+
+  public void stop()
+  {
+    log.info("Stopping scheduled batch scheduler.");
+    final Optional<TaskRunner> taskRunnerOptional = taskMaster.getTaskRunner();
+    if (taskRunnerOptional.isPresent()) {
+      taskRunnerOptional.get().unregisterListener(taskRunnerListener.getListenerId());
+    }
+
+    supervisorToManager.clear();
+  }
+
+  @Nullable
+  public ScheduledBatchSupervisorSnapshot getSchedulerSnapshot(final String supervisorId)
+  {
+    final SchedulerManager manager = supervisorToManager.get(supervisorId);
+    if (manager == null) {
+      return null;
+    }
+
+    final ScheduledBatchStatusTracker.BatchSupervisorTaskStatus tasks = statusTracker.getSupervisorTasks(supervisorId);
+    return new ScheduledBatchSupervisorSnapshot(
+        supervisorId,
+        manager.getSchedulerStatus(),
+        manager.getLastTaskSubmittedTime(),
+        manager.getNextTaskSubmissionTime(),
+        manager.getTimeUntilNextTaskSubmission(),
+        tasks.getSubmittedTasks(),
+        tasks.getCompletedTasks()
```

Comment:
It would probably be better to truncate these lists in the supervisor report and return only the total counts and just a subset of the latest tasks. Otherwise, this could slow down the supervisor status API.

Since these are going to be MSQ tasks, they would have randomly generated group IDs. I wonder if there could be a way to find all the tasks for a given batch supervisor. But at least we can always filter by data source and task type.

Reply:
Good point on tracking the totals and subset. I've made that change by tracking the total counts and only the recent tasks (similar to the compaction supervisor).

>Since these are going to be MSQ tasks, they would have randomly generated group IDs. I wonder if there could be a way to find all the tasks for a given batch supervisor. But at least we can always filter by data source and task type.

Yeah, I initially explored setting the batch supervisor ID as the group ID for MSQ tasks via the query context so users can filter by the supervisor ID. I don’t exactly recall why that didn’t work, but it’s probably worth exploring sometime in the future...